Spigot-Server-Patches/0622-Strip-raytracing-for-EntityLiving-hasLineOfSight.patch

```diff
@@ -0,0 +1,136 @@
+     }
+ 
+     @Override
+diff --git a/src/main/java/net/minecraft/server/IBlockAccess.java b/src/main/java/net/minecraft/server/IBlockAccess.java
+index c4a83448ed4513f6e4ab179d1d43e5bb0cb13641..62258e5d75d259664c107d78d12e7284830bbca2 100644
+--- a/src/main/java/net/minecraft/server/IBlockAccess.java
++++ b/src/main/java/net/minecraft/server/IBlockAccess.java
+@@ -44,6 +44,16 @@ public interface IBlockAccess {
+         return BlockPosition.a(axisalignedbb).map(this::getType);
+     }
+ 
++    // Paper start - broken down variant of below rayTraceBlock, used by World#rayTraceDirect
++    default MovingObjectPosition.EnumMovingObjectType rayTraceBlockDirect(Vec3D vec3d, Vec3D vec3d1, BlockPosition blockposition, IBlockData iblockdata, VoxelShapeCollision voxelshapecoll) {
++        RayTrace.BlockCollisionOption raytrace_blockcollisionoption = RayTrace.BlockCollisionOption.COLLIDER;
++        VoxelShape voxelshape = raytrace_blockcollisionoption.get(iblockdata, this, blockposition, voxelshapecoll); // Collider doesn't actually need a voxelshape
++        MovingObjectPositionBlock movingobjectpositionblock = this.rayTrace(vec3d, vec3d1, blockposition, voxelshape, iblockdata);
++
++        return movingobjectpositionblock == null ? null : movingobjectpositionblock.getType(); // can't be null because of the logic above
++    }
++    // Paper end
++
+     // CraftBukkit start - moved block handling into separate method for use by Block#rayTrace
+     default MovingObjectPositionBlock rayTraceBlock(RayTrace raytrace1, BlockPosition blockposition) {
+             // Paper start - Prevent raytrace from loading chunks
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index 9ed21f434c5fb019b74dfe9ee0b802ccc5c07fd8..f00c8a105c6008a34b0082bfef2826051f8a6451 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -256,6 +256,83 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
+         return null;
+     }
+ 
++    // Paper start - broken down method of raytracing for EntityLiving#hasLineOfSight, replaces IBlockAccess#rayTrace(RayTrace)
++    protected MovingObjectPosition.EnumMovingObjectType rayTraceDirect(Vec3D vec3d, Vec3D vec3d1, VoxelShapeCollision voxelshapecoll) {
++        // most of this code comes from IBlockAccess#a(RayTrace, BiFunction, Function), but removes the needless functions
++        if (vec3d.equals(vec3d1)) {
++            return MovingObjectPosition.EnumMovingObjectType.MISS;
++        }
++        double d0 = MathHelper.d(-1.0E-7D, vec3d1.x, vec3d.x);
++        double d1 = MathHelper.d(-1.0E-7D, vec3d1.y, vec3d.y);
++        double d2 = MathHelper.d(-1.0E-7D, vec3d1.z, vec3d.z);
++        double d3 = MathHelper.d(-1.0E-7D, vec3d.x, vec3d1.x);
++        double d4 = MathHelper.d(-1.0E-7D, vec3d.y, vec3d1.y);
++        double d5 = MathHelper.d(-1.0E-7D, vec3d.z, vec3d1.z);
```

Comment:
OBFHELPER

Reply:
I'm not the submitter and don't plan on changing the patch for this PR, but the reasoning behind not using an obfhelper in these methods is because they're basically hand generated versions of the previous methods. You'll basically need to rewrite them on version changes anyways, to make sure logic doesn't change. It's not a handwritten reimplementation or anything.

Reply:
Rewrites are made easier with obfhelpers either way. When we add new stuff, we want it to be as readable as we can make it.

Reply:
It's complicated here though, because it's not "new stuff". This new method is a stripped version of the other method, so whenever that method changes you have to change this one anyways. I do suppose it helps further optimization of the method, but given you basically have to rewrite this to figure out when the original version of the method changes anyways..